### ISSUE TITLE
librocksdbjni 5.3.4 undefined symbol: ZSTD_versionNumber

### ISSUE THREAD

**Issue by jamespedwards42** [facebook/rocksdb#2280]
I'm getting the following error message since upgrading from version 5.2.1 to 5.3.4 of rocksdbjni.  rocksdbjni  5.3.4 works fine locally on Mac0S with RocksDB 5.3.4 installed through brew. Running it against [RocksDB 5.3.4 on Debian Sid](https://hub.docker.com/r/comodal/debian-rocksdb/~/dockerfile/) is where I get the error. 

`java: symbol lookup error: /tmp/librocksdbjni6228543948847813430.so: undefined symbol: ZSTD_versionNumber`

I moved back to rocksdbjni 5.2.1, keeping everything else the same and it works again.

**Comment by jamespedwards42** [comment]
@adamretter Looks like your recent PR, https://github.com/facebook/rocksdb/pull/2279, will fix this?  I will try to run against it as soon as its out to confirm everything is working.

**Comment by adamretter** [comment]
@jamespedwards42 This shouldn't be happening regardless as when we build the native artifacts, they should be statically linked against their libraries. I will need to try and reproduce this, it will take me some time to setup a Debian environment...

**Comment by nihathrael** [comment]
I have a very similar/the same error using Arch Linux:
```/usr/lib/jvm/java-8-jdk/bin/java: symbol lookup error: /tmp/librocksdbjni7451032689798428193.so: undefined symbol: ZSTD_versionNumber```

**Comment by adamretter** [comment]
@jamespedwards42 Okay so I think I misunderstood what you were attempting to do.

Previously I thought you were using the Java artifacts that we release to Maven Central, but I see from your Dockerfile that you actually are building the tag from source code, and your Makefile target is `shared_lib`. What this implies to me is that when you run RocksJava, it cannot find the dynamically linked library for `libzstd`. I think you need to make sure that that is available on your `LD_LIBRARY_PATH` or similar. 

I will follow the steps of your Dockerfile to try and reproduce on a Debian VM here.

**Comment by nihathrael** [comment]
Hi, I am using the jar file from Maven Central and get the aforementioned error, without using a custom build.

**Comment by adamretter** [comment]
@nihathrael okay I will take a look into that too

**Comment by adamretter** [comment]
@nihathrael @jamespedwards42 I don't understand exactly how the linking works as I am more of a Java than a C++ guy, but I have done some experiments and I believe that https://github.com/facebook/rocksdb/pull/2292 now fixes both of your issues. 

Let me know, and if it works, we will try and cut another release ASAP.

**Comment by jamespedwards42** [comment]
@adamretter I am in the same boat regarding Java versus C++.  I've never built the rocksdb jni library before; I just use what is available in the maven repo.  I will look into building it early this next week and give it a go.

**Comment by adamretter** [comment]
You can just run "make rocksdbjavastatic"

**Comment by jamespedwards42** [comment]
@adamretter Looks like it fixed it!

I checked out your branch, `adamretter:java-zstd-linking`, and built with the following:
```
docker run --rm -it -v /$(pwd)/:/rocksdb -w /rocksdb openjdk:9-jdk bash
apt-get install build-essential
make rocksdbjavastatic
```

Currently running with that jar, with no issues.

**Comment by adamretter** [comment]
@jamespedwards42 Okay thanks for the feedback James. I will try and cut a new release once @sagar0 has confirmed he is happy and has merged that in.

**Comment by nihathrael** [comment]
@adamretter I compiled your branch and everything works as expected. Thanks for the quick fix, much appreciated!

**Comment by adamretter** [comment]
We just released 5.3.6 which I believe now addresses this.
